File: app/api/v1/schemas/documents.py
"""文档相关 Schema。"""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, Field

from app.api.v1.schemas.common import RequestIdMixin
from app.api.v1.schemas.ingest import IngestJobResponse


class DocumentResponse(BaseModel):
    """文档基础信息。"""

    doc_id: str = Field(description="文档ID")
    kb_id: str = Field(description="知识库ID")
    doc_name: str = Field(description="文档名称")
    doc_version: str | None = Field(default=None, description="文档版本")
    published_at: str | None = Field(default=None, description="发布日期")
    source_uri: str | None = Field(default=None, description="文档官方来源链接")
    status: Literal["pending", "processing", "indexed", "failed", "deleted"] = Field(
        description="文档状态"
    )
    error_message: str | None = Field(default=None, description="错误信息")
    chunk_count: int = Field(description="分块数量")
    created_at: str = Field(description="创建时间")
    updated_at: str = Field(description="更新时间")


class DocumentDetailResponse(DocumentResponse, RequestIdMixin):
    """文档详情响应。"""


class DocumentListResponse(RequestIdMixin):
    """文档列表响应。"""

    items: list[DocumentResponse] = Field(description="文档列表")


class DocumentUploadResponse(RequestIdMixin):
    """上传文档响应。"""

    doc: DocumentResponse = Field(description="文档信息")
    job: IngestJobResponse = Field(description="任务信息")


class StagedAssetResponse(BaseModel):
    """暂存图片资产响应。"""

    asset_id: str = Field(description="图片资产 ID")
    label: str = Field(description="图片展示编号")
    file_name: str = Field(description="原始图片文件名")
    media_type: str = Field(description="媒体类型")
    url: str = Field(description="图片访问地址")
    order_index: int = Field(description="图片在文档中的顺序")
    source: str = Field(description="资产来源")
    page_number: int | None = Field(default=None, description="PDF 图片所属页码")


class StagedAssetRefResponse(BaseModel):
    """暂存分块关联的图片资产引用。"""

    asset_id: str = Field(description="图片资产 ID")
    asset_label: str = Field(description="图片展示编号")
    asset_url: str = Field(description="图片访问地址")
    media_type: str | None = Field(default=None, description="媒体类型")
    file_name: str | None = Field(default=None, description="原始图片文件名")


class StagedPageResponse(BaseModel):
    """暂存解析页面响应。"""

    page_number: int | None = Field(default=None, description="页码")
    text: str = Field(description="解析文本")
    section_path: str | None = Field(default=None, description="章节路径")


class StagedChunkResponse(BaseModel):
    """暂存分块响应。"""

    chunk_id: str = Field(description="预览分块 ID")
    chunk_index: int = Field(description="分块序号")
    text: str = Field(description="分块文本")
    page_start: int | None = Field(default=None, description="起始页码")
    page_end: int | None = Field(default=None, description="结束页码")
    section_path: str | None = Field(default=None, description="章节路径")
    enabled: bool = Field(description="是否确认入库")
    source_kind: str = Field(description="分块来源类型")
    asset_id: str | None = Field(default=None, description="关联图片资产 ID")
    asset_label: str | None = Field(default=None, description="关联图片展示编号")
    asset_url: str | None = Field(default=None, description="关联图片访问地址")
    assets: list[StagedAssetRefResponse] | None = Field(
        default=None,
        description="关联图片资产列表",
    )


class StagedPreviewBlockResponse(BaseModel):
    """暂存预览结构块响应。"""

    block_type: Literal["heading", "paragraph", "table", "image"] = Field(
        description="结构块类型"
    )
    order_index: int = Field(description="结构块顺序")
    text: str | None = Field(default=None, description="文本内容")
    level: int | None = Field(default=None, description="标题级别")
    rows: list[list[str]] | None = Field(default=None, description="表格行")
    page_number: int | None = Field(default=None, description="页码")
    section_path: str | None = Field(default=None, description="章节路径")
    asset_id: str | None = Field(default=None, description="关联图片资产 ID")
    asset_label: str | None = Field(default=None, description="关联图片展示编号")
    asset_url: str | None = Field(default=None, description="关联图片访问地址")


class StagedDocumentResponse(RequestIdMixin):
    """暂存文档预览响应。"""

    staged_doc_id: str = Field(description="暂存文档 ID")
    kb_id: str = Field(description="知识库 ID")
    doc_name: str = Field(description="文档名称")
    doc_version: str | None = Field(default=None, description="文档版本")
    published_at: str | None = Field(default=None, description="发布日期")
    source_uri: str | None = Field(default=None, description="文档官方来源链接")
    filename: str = Field(description="原始文件名")
    extension: str = Field(description="扩展名")
    source_type: str = Field(description="来源类型")
    status: Literal["uploaded", "previewed"] = Field(description="暂存状态")
    assets: list[StagedAssetResponse] = Field(description="图片资产列表")
    pages: list[StagedPageResponse] = Field(description="解析页面列表")
    preview_blocks: list[StagedPreviewBlockResponse] = Field(description="版式预览结构块")
    chunks: list[StagedChunkResponse] = Field(description="预览分块列表")
    warnings: list[str] = Field(description="质量提醒")
    created_at: str = Field(description="创建时间")
    updated_at: str = Field(description="更新时间")


class StagedChunkUpdateRequest(BaseModel):
    """暂存分块更新请求。"""

    enabled: bool | None = Field(default=None, description="是否入库")
    text: str | None = Field(default=None, description="分块文本")

File: frontend/src/shared/api/modules/documents.ts
import { apiClient } from "../client";

export interface DocumentItem {
  doc_id: string;
  kb_id: string;
  doc_name: string;
  doc_version?: string | null;
  published_at?: string | null;
  source_uri?: string | null;
  status: "pending" | "processing" | "indexed" | "failed" | "deleted";
  error_message?: string | null;
  chunk_count: number;
  created_at: string;
  updated_at: string;
}

export interface DocumentListResponse {
  items: DocumentItem[];
  request_id?: string | null;
}

export interface IngestProgress {
  stage: string;
  pages_parsed: number;
  chunks_built: number;
  embeddings_done: number;
  vectors_upserted: number;
  stage_ms: number;
  parse_ms: number;
  chunk_ms: number;
  embed_ms: number;
  upsert_ms: number;
}

export interface IngestJob {
  job_id: string;
  kb_id: string;
  doc_id: string;
  status: "queued" | "running" | "succeeded" | "failed" | "canceled";
  progress?: IngestProgress | null;
  error_message?: string | null;
  error_code?: string | null;
  started_at?: string | null;
  finished_at?: string | null;
  created_at: string;
  updated_at: string;
  request_id?: string | null;
}

export interface DocumentUploadResponse {
  doc: DocumentItem;
  job: IngestJob;
  request_id?: string | null;
}

export interface StagedAsset {
  asset_id: string;
  label: string;
  file_name: string;
  media_type: string;
  url: string;
  order_index: number;
  source: string;
}

export interface StagedPage {
  page_number?: number | null;
  text: string;
  section_path?: string | null;
}

export interface StagedChunk {
  chunk_id: string;
  chunk_index: number;
  text: string;
  page_start?: number | null;
  page_end?: number | null;
  section_path?: string | null;
  enabled: boolean;
  source_kind: string;
  asset_id?: string | null;
  asset_label?: string | null;
  asset_url?: string | null;
}

export interface StagedPreviewBlock {
  block_type: "heading" | "paragraph" | "table" | "image";
  order_index: number;
  text?: string | null;
  level?: number | null;
  rows?: string[][] | null;
  page_number?: number | null;
  section_path?: string | null;
  asset_id?: string | null;
  asset_label?: string | null;
  asset_url?: string | null;
}

export interface StagedDocument {
  staged_doc_id: string;
  kb_id: string;
  doc_name: string;
  doc_version?: string | null;
  published_at?: string | null;
  source_uri?: string | null;
  filename: string;
  extension: string;
  source_type: string;
  status: "uploaded" | "previewed";
  assets: StagedAsset[];
  pages: StagedPage[];
  preview_blocks: StagedPreviewBlock[];
  chunks: StagedChunk[];
  warnings: string[];
  created_at: string;
  updated_at: string;
  request_id?: string | null;
}

export async function fetchDocuments(kbId: string) {
  const { data } = await apiClient.get<DocumentListResponse>(`/kb/${kbId}/documents`);
  return data;
}

export async function uploadDocument(params: {
  kbId: string;
  file: File;
  docName?: string;
  docVersion?: string;
  publishedAt?: string;
  sourceUri?: string;
}) {
  const formData = new FormData();
  formData.append("file", params.file);
  if (params.docName) {
    formData.append("doc_name", params.docName);
  }
  if (params.docVersion) {
    formData.append("doc_version", params.docVersion);
  }
  if (params.publishedAt) {
    formData.append("published_at", params.publishedAt);
  }
  if (params.sourceUri) {
    formData.append("source_uri", params.sourceUri);
  }
  const { data } = await apiClient.post<DocumentUploadResponse>(
    `/kb/${params.kbId}/documents`,
    formData
  );
  return data;
}

export async function uploadStagedDocument(params: {
  kbId: string;
  file: File;
  docName?: string;
  docVersion?: string;
  publishedAt?: string;
  sourceUri?: string;
}) {
  const formData = new FormData();
  formData.append("file", params.file);
  if (params.docName) {
    formData.append("doc_name", params.docName);
  }
  if (params.docVersion) {
    formData.append("doc_version", params.docVersion);
  }
  if (params.publishedAt) {
    formData.append("published_at", params.publishedAt);
  }
  if (params.sourceUri) {
    formData.append("source_uri", params.sourceUri);
  }
  const { data } = await apiClient.post<StagedDocument>(
    `/kb/${params.kbId}/documents/staged`,
    formData
  );
  return data;
}

export async function buildStagedPreview(stagedDocId: string) {
  const { data } = await apiClient.post<StagedDocument>(
    `/staged-documents/${stagedDocId}/preview`
  );
  return data;
}

export async function updateStagedChunk(params: {
  stagedDocId: string;
  chunkId: string;
  enabled?: boolean;
  text?: string;
}) {
  const { data } = await apiClient.patch<StagedDocument>(
    `/staged-documents/${params.stagedDocId}/chunks/${params.chunkId}`,
    {
      enabled: params.enabled,
      text: params.text
    }
  );
  return data;
}

export async function commitStagedDocument(stagedDocId: string) {
  const { data } = await apiClient.post<DocumentUploadResponse>(
    `/staged-documents/${stagedDocId}/commit`
  );
  return data;
}

export async function deleteDocument(docId: string) {
  const { data } = await apiClient.delete<{ status: string; request_id?: string | null }>(
    `/documents/${docId}`
  );
  return data;
}

export async function reindexDocument(docId: string) {
  const { data } = await apiClient.post<IngestJob>(`/documents/${docId}/reindex`);
  return data;
}

export async function fetchIngestJob(jobId: string) {
  const { data } = await apiClient.get<IngestJob>(`/ingest/jobs/${jobId}`);
  return data;
}

export async function cancelIngestJob(jobId: string) {
  const { data } = await apiClient.post<IngestJob>(`/ingest/jobs/${jobId}/cancel`);
  return data;
}

export async function retryIngestJob(jobId: string) {
  const { data } = await apiClient.post<IngestJob>(`/ingest/jobs/${jobId}/retry`);
  return data;
}
